src/main/scala/org/broadinstitute/dsde/firecloud/EntityClient.scala

```diff
@@ -253,52 +253,97 @@ class EntityClient (requestContext: RequestContext) extends Actor {
     EntityUpdateDefinition(row.headOption.get,entityType,ops ++ collectionMemberAttrOp )
   }
 
-  def addToCollectionTypeFromTSV( entityType: String, memberType: String, membersAttributeName: String, tsv: TSVLoadFile ) = {
-    // group together updates to distinct entities
-    tsv.tsvData groupBy(_(0)) map { case (entityName, rows) =>
-      val ops = rows map { row =>
-        val attrRef = AttributeReference(memberType,row(1)) //row(1) is the entity to add to the entity in row.head
-        Map(addListMemberOperation,"attributeListName"->AttributeString(membersAttributeName),"newMember"->attrRef)
-      }
-      EntityUpdateDefinition(entityName,entityType,ops)
+  private def validateMembershipTSV(tsv: TSVLoadFile, membersType: Option[String]) (op: => Unit): Unit = {
+    //This magical list of conditions determines whether the TSV is populating the "members" attribute of a collection type entity.
+    if( !membersType.isDefined ) {
+      requestContext.complete( HttpResponse(BadRequest, "Invalid membership TSV. Entity type must be a collection type") )
+    } else if( tsv.headers.length != 2 ){
+      requestContext.complete( HttpResponse(BadRequest, "Invalid membership TSV. Must have exactly two columns") )
+    } else if( tsv.headers != Seq(tsv.firstColumnHeader, membersType.get + "_id") ) {
+      requestContext.complete( HttpResponse(BadRequest, "Invalid membership TSV. Second column header should be " + membersType.get + "_id") )
+    } else {
+      op
```

Comment:
i bet @geoffjentry knows a monad that knows an idiomat that knows a more scalay way to do this than ifs / else { op }, but ill probably wish i never said that after i see the result